fix: use validated schema to ensure defaults are passed on insert

DawidWraga opened the pull request.

KyleAMathews commented:
Hey it's looking good! A few failing tests to fix still...

KyleAMathews commented:
@samwillis could you review my type changes? Particularly to the query builder? Don't want to mess with any guarantees you wanted. Our types are getting complicated... :-\

DawidWraga commented:
ah also I forgot to update the collection jsdoc string `@templates` :

```ts
// packages/db/src/collection.ts

/**
 * Enhanced Collection interface that includes both data type T and utilities TUtils
 * @template T - The type of items in the collection
 * @template TKey - The type of the key for the collection
 * @template TUtils - The utilities record type
 * @template TSchema - The schema type for validation and type inference
 */
export interface Collection<

```

KyleAMathews commented:
Hmm DB doesn't have the concept of derived fields atm that the dev can't modify directly. Our general assumption is you'd calculate those in `select` in queries & share query fragments as needed. There is benefit to putting the derived field directly in the collection for sure to be easily reusable so we might want to do it in the future but it does complicate the collection implementation enough that my preference is to not do it unless there's a lot of interest.

samwillis commented:
> @samwillis could you review my type changes? Particularly to the query builder? Don't want to mess with any guarantees you wanted. Our types are getting complicated... :-\

All the code in this file is completely re-written in the new query builder, so there is little point in making this perfect if its TTL is a better of days...

We will need to make similar changes in the new query builder - I think they looks fine, although I haven't dug into what the new generic args are for.

KyleAMathews commented on packages/db/tests/collection.test.ts:
```diff
@@ -833,4 +833,93 @@ describe(`Collection with schema validation`, () => {
+
+    // Minimal data
+    const tx1 = createTransaction({ mutationFn })
+    tx1.mutate(() => collection.insert({ text: `task-1` }))
```
can you test asserting on tx.mutations[0] (changes, etc.)?

KyleAMathews replied:
also add some type tests in the collection.test-d.ts file

DawidWraga commented:
Hey @KyleAMathews  I've written some tests but having trouble deciding what to do with the mutation types 

If we want full type safety on the `PendingMutation` - including the changes type being different for delete/update/insert based on the TOperation, then it adds a significant amount of complexity for a little bit more type safety, I'm not sure whether the trade off is worth it.


All of these helpers:
```ts

export type UpdateMutationFnParams<T extends object = Record<string, unknown>> =
  {
    transaction: TransactionWithMutations<T>
  }

export type InsertMutationFnParams<T extends object = Record<string, unknown>> =
  {
    transaction: TransactionWithMutations<T>
  }

export type DeleteMutationFnParams<T extends object = Record<string, unknown>> =
  {
    transaction: TransactionWithMutations<T>
  }

export type InsertMutationFn<T extends object = Record<string, unknown>> = (
  params: InsertMutationFnParams<T>
) => Promise<any>

export type UpdateMutationFn<T extends object = Record<string, unknown>> = (
  params: UpdateMutationFnParams<T>
) => Promise<any>

export type DeleteMutationFn<T extends object = Record<string, unknown>> = (
  params: DeleteMutationFnParams<T>
) => Promise<any>

// ...

/**
 * Utility type for a Transaction with at least one mutation
 * This is used internally by the Transaction.commit method
 */
export type TransactionWithMutations<
  T extends object = Record<string, unknown>,
> = Transaction<T> & {
  mutations: NonEmptyArray<PendingMutation<T>>
}

export interface TransactionConfig<T extends object = Record<string, unknown>> {
  /** Unique identifier for the transaction */
  id?: string
  /* If the transaction should autocommit after a mutate call or should commit be called explicitly */
  autoCommit?: boolean
  mutationFn: MutationFn<T>
  /** Custom metadata to associate with the transaction */
  metadata?: Record<string, unknown>
}



```

(and more)


All these helpers ultimately wrap around `PendingMutation`. This is what it looks like in main right now:

```ts

export interface PendingMutation<
  T extends object = Record<string, unknown>,
  TOperation extends OperationType = OperationType,
> {
 // ...
  original: TOperation extends `insert` ? {} : T
  // The result state of the object after all mutations.
  modified: T
  // The actual changes made.
  changes: TOperation extends `insert`
    ? T
    : TOperation extends `delete`
      ? T
      : Partial<T>
      
   // without TInsertInput
    collection: Collection<T, any>
   //...
      }
```


So if you wanted to make the changes property be fully type safe you could do something like this:

```ts

export type ResolveTransactionData<
  T extends object = Record<string, unknown>,
  TOperation extends OperationType = OperationType,
  TInsertInput extends object = T,
> = TOperation extends `insert`
  ? TInsertInput
  : TOperation extends `delete`
    ? T
    : Partial<T>

export interface PendingMutation<
  T extends object = Record<string, unknown>,
  TOperation extends OperationType = OperationType,
  TInsertInput extends object = T,
  TCollection extends Collection<T, any, any, any, TInsertInput> = Collection<
    T,
    any,
    any,
    any,
    TInsertInput
  >,
> {
  mutationId: string
  // The state of the object before the mutation.
  original: TOperation extends `insert` ? {} : T
  // The result state of the object after all mutations.
  modified: T
  // The actual changes made.
  changes: ResolveTransactionData<T, TOperation, TInsertInput>

  collection: TCollection
  
  //....
}

```

But then the trouble is - you would have to pass many generics to all the mutation helpers. 


I tried to refactor following this pattern to minimize complexity:
Inside `createCollection` we simplify `TExplicit`, `TSchema` and `TFallback` on `Collection`  into just `TInput`
Similarly, inside `createTransaction` I tried to simplify `T`, `TOperation` and `TInsertInput` into just `T` 

However, unlike `Collection`, `PendingMutations`  explicitly requires all 3 to be type safe (`T` for modified , `TOperation` for type and `TInsertInput` for changes). If we just relied on the generic defaults then it kinda defeats the point - no benefit but adds lots of extra code.


On collection the complexity is lower and definitely worth it to make the operation params fully type safe. 
However, on the mutations I'm not sure if it's worth the complexity trade off.


It could just be left as `TOperation extends `delete` ? T  : Partial<T>` like it was before to keep things simple.


Not sure, it's up to you. I'll push my attempt for visibility but it's just a draft currently.


Current state of the code:
- I've added some tests
- Initially, added the new generics to most the type utils (most tests passing, but led to lots of extra types)
- Tried to simplify using `ResolveTransactionData` inside `createTransaction` but then realised `PendingMutation` requires all 3.
- Uncertain about trade-off between complexity and typesafety, stopped development, wrote this post instead

You're welcome to pick it up from here if you like, or I can try something more later this week, my brain is pretty fried for a moment from reading all those types 😆

KyleAMathews commented:
This might be beyond me 😅 @kevin-dp any thoughts?

kevin-dp commented:
I've been chatting this through with @KyleAMathews. Here are my thoughts:
- i agree that we want to be able to omit columns that have defaults when inserting
- regarding the `changes` i think it's fine to type them as `Partial<T>` which keeps the code simpler and more maintainable at the cost of a slightly less precise type.

Having precise types for user **input** is really important because e.g. when developers pass an argument to a function and the argument type checks, you want the function to work on that input. That's why `Partial<T>` is not acceptable for the argument to `insert` because if the user would omit a required field it will typecheck but won't work at runtime.

However, the type of **outputs** is a different story. When the output of a function (or the `changes` field in our example) is imprecise like `Partial<T>` that is still correct because all of the changes will fulfill that type. So if the developer's code can handle changes of type `Partial<T>` (i.e. the code typechecks) then at runtime it will be able to handle any of the actual changes it gets (because all changes are of type `Partial<T>` even if that's not the most precise type for it).

I don't think the slightly improved typing for `changes` is worth the added complexity. In essence, the difference with your suggested approach would be that instead of typing the changes as `Partial<T>` (which makes every column optional) you would only make the columns that have defaults optional.

DawidWraga commented:
Thanks for your input @kevin-dp  that's exactly what I was thinking, completely agree.

I'll finish this up now 😄

KyleAMathews commented:
Actually did another through go over and it looks good to go — please do check though @DawidWraga when you have a chance and PR anything I missed!

KyleAMathews commented:
Hmm good question. I think c still makes the most sense as "changes" are how the model was changed not the raw inputs. I'd find it odd if people wouldn't want the cleaned up data. If they didn't, they should probably change their schema? I'm not super confident or feel strongly about this. I suspect either way would be fine but keeping everything at the end state sounds good right now.

marbemac commented:
Maybe I'm not understanding the change here, but should the below work?

```ts
const todoSchema = z.object({
  name: z.string(),

  // createdAt should not be required on insert
  createdAt: z.date().default(() => new Date()),
})

const todosCollection = createCollection(
  queryCollectionOptions({
    id: 'todos',
    queryKey: ['todos'],
    queryFn: async () => {
      return [
        { name: 'Todo 1', createdAt: new Date() },
        { name: 'Todo 2', createdAt: new Date() },
      ];
    },
    queryClient,
    schema: todoSchema,
    getKey: item => item.name,
    onInsert: async ({ transaction }) => {
      const { modified: newTodo } = transaction.mutations[0];
    },
  }),
);

// This is erroring
todosCollection.insert({ name: 'new todo' });
```

It seems as though `insert` is typed off of the return type from `queryFn`. This is using latest versions, and `@tanstack/query-db-collection`.

<img width="1287" height="367" alt="Screenshot 2025-08-30 at 5 18 04 PM" src="https://github.com/user-attachments/assets/a13cff89-1fe0-4553-9155-eddcdfeeb896" />
